File: client_app/apple_shared/Rpt2/Sources/Rpt2/RptTrafficShape.swift
import Foundation

/// Product traffic-shape helpers — mirrors `node/traffic_shape.py`.
///
/// Wire (AEAD plaintext):
///   Real padded: RPTP || u16_be(len) || plain || random_pad
///   Cover dummy: RPTC || random_bytes
public enum RptTrafficShape {
    public static let padMagic = Data("RPTP".utf8)
    public static let coverMagic = Data("RPTC".utf8)
    public static let productPadBucket: Int = 128
    public static let productCoverSize: Int = 128
    public static let productCoverIntervalS: TimeInterval = 2.0
    public static let productPadding: Bool = true
    public static let productCover: Bool = true

    public static func padPayload(_ plain: Data, bucket: Int = productPadBucket) throws -> Data {
        guard plain.count <= 65535 else {
            throw RptProtocol.ProtocolError("plain too large for u16 length")
        }
        let b = max(16, min(2048, bucket))
        var lenBe = UInt16(plain.count).bigEndian
        var body = Data(bytes: &lenBe, count: 2)
        body.append(plain)
        let padLen = (b - (body.count % b)) % b
        var out = padMagic
        out.append(body)
        if padLen > 0 {
            var pad = Data(count: padLen)
            _ = pad.withUnsafeMutableBytes {
                SecRandomCopyBytes(kSecRandomDefault, padLen, $0.baseAddress!)
            }
            out.append(pad)
        }
        return out
    }

    public static func makeCoverPayload(size: Int = productCoverSize) -> Data {
        let n = max(16, min(2048, size))
        var noise = Data(count: n - coverMagic.count)
        _ = noise.withUnsafeMutableBytes {
            SecRandomCopyBytes(kSecRandomDefault, noise.count, $0.baseAddress!)
        }
        var out = coverMagic
        out.append(noise)
        return out
    }

    /// Returns (payload, isCover). Unmarked blobs treated as raw IP (compat).
    public static func unpadPayload(_ blob: Data) throws -> (Data, Bool) {
        if blob.count >= 4, blob.prefix(4) == coverMagic {
            return (Data(), true)
        }
        if blob.count < 4 || blob.prefix(4) != padMagic {
            return (blob, false)
        }
        let rest = blob.dropFirst(4)
        guard rest.count >= 2 else {
            throw RptProtocol.ProtocolError("truncated padded payload")
        }
        let n = Int(rest[rest.startIndex]) << 8 | Int(rest[rest.startIndex + 1])
        guard rest.count >= 2 + n else {
            throw RptProtocol.ProtocolError("padded payload length exceeds buffer")
        }
        let start = rest.index(rest.startIndex, offsetBy: 2)
        let end = rest.index(start, offsetBy: n)
        return (Data(rest[start..<end]), false)
    }

    public static func prepareOutbound(_ ipPacket: Data, padding: Bool = productPadding) throws -> Data {
        if padding { return try padPayload(ipPacket) }
        return ipPacket
    }

    public static func interpretInbound(_ blob: Data) throws -> (Data?, Bool) {
        let (plain, isCover) = try unpadPayload(blob)
        if isCover { return (nil, true) }
        return (plain, false)
    }
}

File: client_app/android/app/src/main/kotlin/com/restoreprivacy/restore_privacy_client/RptTrafficShape.kt
package com.restoreprivacy.restore_privacy_client

import java.nio.ByteBuffer
import java.security.SecureRandom

/**
 * Product traffic-shape helpers — mirrors `node/traffic_shape.py`.
 *
 * Wire (AEAD plaintext):
 *   Real padded: RPTP || u16_be(len) || plain || random_pad
 *   Cover dummy: RPTC || random_bytes
 *
 * Product residual defaults: padding on (bucket 128), cover available.
 */
object RptTrafficShape {
    val PAD_MAGIC: ByteArray = byteArrayOf(0x52, 0x50, 0x54, 0x50) // RPTP
    val COVER_MAGIC: ByteArray = byteArrayOf(0x52, 0x50, 0x54, 0x43) // RPTC
    const val PRODUCT_PAD_BUCKET: Int = 128
    const val PRODUCT_COVER_SIZE: Int = 128
    const val PRODUCT_COVER_INTERVAL_MS: Long = 2000L
    /** Bounded send-side delay (ms); matches Python product policy (jitter_ms_max=40). */
    const val PRODUCT_JITTER_MS_MAX: Int = 40
    const val PRODUCT_PADDING: Boolean = true
    const val PRODUCT_COVER: Boolean = true

    private val rnd = SecureRandom()

    /** Product residual send jitter (0…PRODUCT_JITTER_MS_MAX). DATA path only. */
    fun applySendJitter() {
        val maxMs = PRODUCT_JITTER_MS_MAX
        if (maxMs <= 0) return
        val ms = rnd.nextInt(maxMs + 1)
        if (ms > 0) {
            try {
                Thread.sleep(ms.toLong())
            } catch (_: InterruptedException) {
                Thread.currentThread().interrupt()
            }
        }
    }

    fun padPayload(plain: ByteArray, bucket: Int = PRODUCT_PAD_BUCKET): ByteArray {
        require(plain.size in 0..65535) { "plain too large for u16 length" }
        val b = bucket.coerceIn(16, 2048)
        val lenBe = ByteBuffer.allocate(2).putShort(plain.size.toShort()).array()
        val body = lenBe + plain
        val padLen = (b - (body.size % b)) % b
        val pad = if (padLen > 0) ByteArray(padLen).also { rnd.nextBytes(it) } else ByteArray(0)
        return PAD_MAGIC + body + pad
    }

    fun makeCoverPayload(size: Int = PRODUCT_COVER_SIZE): ByteArray {
        val n = size.coerceIn(16, 2048)
        val noise = ByteArray(n - COVER_MAGIC.size).also { rnd.nextBytes(it) }
        return COVER_MAGIC + noise
    }

    /** Returns (payload, isCover). Unmarked blobs treated as raw IP (compat). */
    fun unpadPayload(blob: ByteArray): Pair<ByteArray, Boolean> {
        if (blob.size >= 4 && blob.copyOfRange(0, 4).contentEquals(COVER_MAGIC)) {
            return ByteArray(0) to true
        }
        if (blob.size < 4 || !blob.copyOfRange(0, 4).contentEquals(PAD_MAGIC)) {
            return blob to false
        }
        val rest = blob.copyOfRange(4, blob.size)
        if (rest.size < 2) throw IllegalArgumentException("truncated padded payload")
        val n = ((rest[0].toInt() and 0xff) shl 8) or (rest[1].toInt() and 0xff)
        if (rest.size < 2 + n) throw IllegalArgumentException("padded payload length exceeds buffer")
        return rest.copyOfRange(2, 2 + n) to false
    }

    fun prepareOutbound(ipPacket: ByteArray, padding: Boolean = PRODUCT_PADDING): ByteArray {
        return if (padding) padPayload(ipPacket) else ipPacket
    }

    fun interpretInbound(blob: ByteArray): Pair<ByteArray?, Boolean> {
        val (plain, isCover) = unpadPayload(blob)
        return if (isCover) null to true else plain to false
    }
}
